### PR TITLE
Added web socket routing

### LINKED ISSUE #80: WebSocketConnection Routing
Implement routing for websocket messages to the correct handlers.

We should probably only implement this for the /v1/messages endpoint right now, but keep in mind that there are many endpoints that can be called this way.

There is already helper functions to extract data from uri's aswell as extracting / unpacking of signal messages from the protobufs. refer to the `managers/websocket/net_helper.rs` file to see more. 

this is blocked by #35

### PR DESCRIPTION
**Pull request by Darkros1245**
Added web socket routing.
We have to figure out how to delete messages after they have been sent.
Closes #80

### CODE REVIEW

**Review by MJHC — COMMENTED** [review]
check if http endpoint still works after changed func signature

**Review by MJHC — APPROVED** [review]
okay

**Review by andreasalstrup — CHANGES_REQUESTED** [review]
test managers::websocket::connection::test::test_on_receive_request ... FAILED

failures:

---- managers::websocket::connection::test::test_on_receive_request stdout ----
thread 'managers::websocket::connection::test::test_on_receive_request' panicked at server/src/managers/websocket/connection.rs:566:14:
called `Result::unwrap()` on an `Err` value: "Send failed"
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    managers::websocket::connection::test::test_on_receive_request

test result: FAILED. 62 passed; 1 failed; 9 ignored; 0 measured; 0 filtered out; finished in 2.19s

**Comment by MJHC** [comment]
no

File: server/src/managers/websocket/connection.rs [hunk 1]
```diff
@@ -177,8 +182,119 @@ impl<W: WSStream + Debug + Send + 'static, DB: SignalDatabase + Send + 'static>
...
+                    msq_id,
+                    StatusCode::INTERNAL_SERVER_ERROR.as_u16() as u32,
+                    StatusCode::INTERNAL_SERVER_ERROR
+                        .canonical_reason()
+                        .ok_or("Invalid canonical reason")?,
+                    vec![],
+                    None,
+                )
+                .encode_to_vec(),
+            ));
```
Comment (andreasalstrup) [inline server/src/managers/websocket/connection.rs:213]: Der mangler?

`.await`

File: server/src/managers/websocket/connection.rs [hunk 2]
```diff
@@ -177,8 +182,120 @@ impl<W: WSStream + Debug + Send + 'static, DB: SignalDatabase + Send + 'static>
...
+        }
+    }
+
+    async fn handle_request(&mut self, request_msq: WebSocketRequestMessage) -> Result<(), String> {
+        let msq_id = request_msq.id.ok_or("Request id was not present")?;
+        if !request_msq.path().starts_with("/v1/messages") {
+            self.send(Message::Binary(
+                create_response(
+                    msq_id,
+                    StatusCode::INTERNAL_SERVER_ERROR.as_u16() as u32,
```
Comment (MJHC) [inline server/src/managers/websocket/connection.rs:205]: only take statuscode one time

File: server/src/managers/websocket/connection.rs [hunk 3]
```diff
@@ -373,10 +533,56 @@ pub(crate) mod test {
...
+                &format!("/v1/messages/{}", client.protocol_address().name()),
+                vec![],
+                Some(msg),
+            ))
+            .await
+            .unwrap();
+    }
+
+    #[tokio::test]
+    async fn test_on_receive_response() {
```
Comment (MJHC) [inline server/src/managers/websocket/connection.rs:572]: add a test that tests that one client can send a message to another, by inserting two clients into the websocketmanager and having alice's sender send a message and then bobs receiver receive a message after a short sleep

**Review by MJHC — APPROVED** [review]
okay

File: server/src/account.rs [hunk 4]
```diff
@@ -103,6 +103,7 @@ pub struct Device {
 }
 #[allow(clippy::too_many_arguments)]
 impl Device {
```
Comment (Laursen79) [inline server/src/account.rs:105]: brug bon?

File: server/src/server.rs [hunk 5]
```diff
@@ -81,7 +81,7 @@ async fn handle_put_messages<T: SignalDatabase, U: WSStream + Debug>(
     )
     .map_err(|_| ApiError {
         status_code: StatusCode::INTERNAL_SERVER_ERROR,
-        message: "".to_owned(),
+        message: "a".to_owned(),
```
Comment (Laursen79) [inline server/src/server.rs:84]: Hvad??

File: server/src/server.rs [hunk 6]
```diff
@@ -90,12 +90,12 @@ async fn handle_put_messages<T: SignalDatabase, U: WSStream + Debug>(
     )
     .map_err(|_| ApiError {
         status_code: StatusCode::INTERNAL_SERVER_ERROR,
-        message: "".to_owned(),
+        message: "b".to_owned(),
```
Comment (Laursen79) [inline server/src/server.rs:93]: Hvad??

File: server/src/server.rs [hunk 7]
```diff
@@ -105,8 +105,9 @@ async fn handle_put_messages<T: SignalDatabase, U: WSStream + Debug>(
             destination.aci().service_id_string(),
             message.destination_device_id.into(),
         );
-        state.message_manager.insert(&address, &mut envelope);
-    });
+        state.message_manager.insert(&address, &mut envelope).await; // TODO: Remember to check for
```
Comment (Laursen79) [inline server/src/server.rs:108]: Todo?

File: server/src/server.rs [hunk 8]
```diff
@@ -237,7 +238,13 @@ async fn put_messages_endpoint(
     Json(payload): Json<SignalMessages>, // TODO: Multiple messages could be sent at one time
```
Comment (Laursen79) [inline server/src/server.rs:238]:
```suggestion
    Json(payload): Json<SignalMessages>,
```

File: server/src/server.rs [hunk 9]
```diff
@@ -347,9 +354,15 @@ async fn create_websocket_endpoint(
...
+                state.clone(),
             );
-            wmgr.insert(ws, receiver).await
+            //let addr = ws.protocol_address();
+            wmgr.insert(ws, receiver).await;
+            //let ws = wmgr.get(&addr).await.unwrap();
+            //state
+            //    .message_manager
+            //    .add_message_availability_listener(&addr, ws)
+            //    .await;
```
Comment (Laursen79) [inline server/src/server.rs:365]: What is this?

**Review by Laursen79 — CHANGES_REQUESTED** [review]

**Review by Laursen79 — APPROVED** [review]

**Review by MJHC — APPROVED** [review]
sticker(1)

**Review by andreasalstrup — APPROVED** [review]

**Pull request merged**